sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedReadFromBoundedSource.java

```diff
@@ -0,0 +1,204 @@
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PInput;
+
+import com.google.api.client.util.Lists;
+import com.google.common.annotations.VisibleForTesting;
+
+import org.joda.time.Instant;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+/**
+ * {@link PTransform} that performs a unbounded read from an {@link BoundedSource}.
+ *
+ * <p>Created by {@link Read}.
+ */
+public class UnboundedReadFromBoundedSource<T> extends PTransform<PInput, PCollection<T>> {
+  private final BoundedSource<T> source;
+
+  UnboundedReadFromBoundedSource(BoundedSource<T> source) {
+    this.source = source;
+  }
+
+  @Override
+  public PCollection<T> apply(PInput input) {
+    return Pipeline.applyTransform(input,
+        Read.from(new BoundedToUnboundedSourceAdapter<>(source)));
+  }
+
+  @Override
+  protected Coder<T> getDefaultOutputCoder() {
+    return source.getDefaultOutputCoder();
+  }
+
+  @Override
+  public String getKindString() {
+    return "Read(" + approximateSimpleName(source.getClass()) + ")";
+  }
+
+  @Override
+  public void populateDisplayData(DisplayData.Builder builder) {
+    // We explicitly do not register base-class data, instead we use the delegate inner source.
+    builder
+        .add(DisplayData.item("source", source.getClass()))
+        .include(source);
+  }
+
+  @VisibleForTesting
+  static class BoundedToUnboundedSourceAdapter<T>
+      extends UnboundedSource<T, BoundedToUnboundedSourceAdapter.Checkpoint> {
+
+    private final BoundedSource<T> boundedSource;
+
+    public BoundedToUnboundedSourceAdapter(BoundedSource<T> boundedSource) {
+      this.boundedSource = boundedSource;
+    }
+
+    @Override
+    public void validate() {
+      boundedSource.validate();
+    }
+
+    @Override
+    public List<BoundedToUnboundedSourceAdapter<T>> generateInitialSplits(
+        int desiredNumSplits, PipelineOptions options) throws Exception {
+      long desiredBundleSize = boundedSource.getEstimatedSizeBytes(options) / desiredNumSplits;
+      List<BoundedToUnboundedSourceAdapter<T>> result = Lists.newArrayList();
+      for (BoundedSource<T> split : boundedSource.splitIntoBundles(desiredBundleSize, options)) {
+        result.add(new BoundedToUnboundedSourceAdapter<>(split));
+      }
+      return result;
+    }
+
+    @Override
+    public Reader createReader(PipelineOptions options, Checkpoint checkpoint) throws IOException {
+      return new Reader(boundedSource.createReader(options), checkpoint);
+    }
+
+    @Override
+    public Coder<T> getDefaultOutputCoder() {
+      return boundedSource.getDefaultOutputCoder();
+    }
+
+    @Override
+    public boolean requiresDeduping() {
```

Comment:
This is incomplete - You must provide an implementation of `getCurrentRecordId()` for `requiresDeduping()` to function (otherwise it will throw an exception in calls to `getCurrentRecordId()`). Probably right to use (element number, shardId) in this invocation. AFAIK, deduping is best-effort (not over the life of a pipeline), so this may produce duplicate elements in some cases.

This also does not solve the problem of progress - the reader may be discarded before it returns `false` to `start()` or `advance()` (as is done in the `InProcessPipelineRunner`), which may cause to the reader producing the same subset of elements and never completing the input. I believe reading the entire contents into an `Iterable` within `start()` and outputting (and then flattening) that `Iterable` will provide completion and progress guarantees, excepting the case in which not all elements fit in memory. Potentially the produced `Iterable` can be implemented as a channel back to the underlying `BoundedReader` (and thus lazily produce elements as `TimestampedValues`, which can be written to some channel and cleared out of memory if supported by the runner).